### PR TITLE
IAP transport read/writes on background thread

### LINKED ISSUE #522: IAPTransport Writes Output Stream from Different Thread than Scheduled Run Loop
The input and output streams for the EASessions used by the SDL mobile proxy are scheduled and opened on the main thread in SDLIAPSession's startStream method. However, the output stream is written to from the dispatch queue com.sdl.transport.iap.transmit in SDLIAPTransport's sendData method. This cross-thread usage of the output stream appears to violate the Apple usage guidelines for NSOutputStream, as stated in https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/Streams/Articles/WritingOutputStreams.html. I have bold formatted the specific guideline from the relevant paragraph.

Before you open the stream to begin the streaming of data, send a scheduleInRunLoop:forMode: message to the stream object to schedule it to receive stream events on a run loop. By doing this, you are helping the delegate to avoid blocking when the stream is unable to accept more bytes. If streaming is taking place on another thread, be sure to schedule the stream object on that thread’s run loop. **You should never attempt to access a scheduled stream from a thread different than the one owning the stream’s run loop**. Finally, send the NSOutputStream instance an open message to start the streaming of data to the output container.

The IAPSession startStream method should be extended to include a dispatch_queue argument that should be set to nil if the main run loop is to be used and to the custom dispatch queue to be used otherwise. Then as discussed in this StackOverflow article, the run loop for the thread backing the dispatch queue should be used and the run loop's run method must be called:

http://stackoverflow.com/questions/14022044/how-to-schedule-event-in-runloop-in-gcd

Alternatively, an NSThread can be detached to process the output stream.

### PR DESCRIPTION
**Pull request by davidswi**
Fixes #522, #424 

This PR is ready for review.

### Risk
This PR makes no API changes.

### Testing Plan
We performed manual testing with a number of internal SDL test apps and prototype headunit hardware, covering validation and verification of correct RPC and a variety of cold and warm connect/disconnect scenarios over BT and USB. We are adding automated testing to perform longhaul characterization of MTTF.

### Summary
This PR moves all IAP transport reads and writes for the active data session to a background thread. Send buffers are enqueued from the IAPTransport sendData method to an internal FIFO managed by the IAPSession instance. They are dequeued and consumed by the background thread using an initial stream write to consume as much of the buffer as possible, and by scheduled writes using the delegate space available callback to consume the remainder as space is available, thereby minimizing CPU utilization and preserving battery life when the iOS device is unplugged. For plugged scenarios, battery charging should be faster due to the minimized CPU utilization.

In addition, race conditions caused by sending events to deallocating proxy objects have been fixed. Crashes and hangs due to these race conditions should now be reduced to rare occurrences.

##### Breaking Changes
* No API breaking changes.

##### Enhancements
* Data session I/O is now in full compliance with Apple guidelines.

##### Bug Fixes
* All session output stream API calls are made on the scheduled thread, scheduled thread is no longer the main thread and polling is no longer used to minimize CPU impact.

### Tasks Remaining:
None. An additional PR for properly wrapping accessory connection in a background task will be opened shortly.

### CODE REVIEW

**Comment by joeljfischer** [comment]
This actually fixes #522, correct? So #566 can be closed?

**Comment by davidswi** [comment]
Hi Joel, yes #566 can be closed. This PR also fixes #424.

File: SmartDeviceLink/SDLMutableDataQueue.m [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    }
+}
+
+- (NSUInteger)count {
+    @synchronized (self) {
+        return self.elements.count;
+    }
+}
+
+- (void)dealloc {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMutableDataQueue.m]: This shouldn't be necessary

File: SmartDeviceLink/SDLMutableDataQueue.m [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        return nil;
+    }
+    
+    self.elements = [[NSMutableArray alloc] init];
+    
+    return self;
+}
+
+- (void)enqueueBuffer:(NSMutableData *)data {
+    @synchronized (self) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLMutableDataQueue.m]: I assume we're synchronizing all of this to make it thread safe. Are we expecting it to be used on multiple threads? If not, synchronization is relatively slow.

File: SmartDeviceLink/SDLProtocol.m [hunk 3]
```diff
@@ -229,7 +229,10 @@ - (BOOL)sendRPC:(SDLRPCMessage *)message encrypted:(BOOL)encryption error:(NSErr
             }
         } break;
         default: {
+            /*
             NSAssert(NO, @"Attempting to send an RPC based on an unknown version number: %@, message: %@", @([SDLGlobals globals].protocolVersion), message);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocol.m]: Why was this commented out? For debug purposes? Asserts only occur in DEBUG.

File: SmartDeviceLink/SDLProtocol.m [hunk 4]
```diff
@@ -347,11 +350,18 @@ - (void)handleBytesFromTransport:(NSData *)receivedData {
 }
 
 - (void)processMessages {
+    SDLProtocolHeader *header;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocol.m]: This should initialize explicitly with `nil` if necessary.

File: SmartDeviceLink/SDLProtocol.m [hunk 5]
```diff
@@ -347,11 +350,18 @@ - (void)handleBytesFromTransport:(NSData *)receivedData {
...
     NSMutableString *logMessage = [[NSMutableString alloc] init];
-    UInt8 incomingVersion = [SDLProtocolMessage determineVersion:self.receiveBuffer];
-
-    // If we have enough bytes, create the header.
-    SDLProtocolHeader *header = [SDLProtocolHeader headerForVersion:incomingVersion];
+    
+    if (self.receiveBuffer != nil && self.receiveBuffer.length > 0){
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocol.m]: This should be flipped and return early.

File: SmartDeviceLink/SDLProtocolHeader.h [hunk 6]
```diff
@@ -37,6 +37,8 @@ typedef NS_ENUM(UInt8, SDLFrameData) {
     SDLFrameData_ConsecutiveLastFrame = 0x00
 };
 
+UInt8 const invalidSDLProtocolVersion = 255;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocolHeader.h]: Should be capitalized. And equal UINT8_MAX instead.

File: SmartDeviceLink/SDLProtocolHeader.m [hunk 7]
```diff
@@ -59,8 +58,12 @@ + (__kindof SDLProtocolHeader *)headerForVersion:(UInt8)version {
             return [[SDLV2ProtocolHeader alloc] initWithVersion:version];
         } break;
         default: {
+#ifdef DEBUG
             NSString *reason = [NSString stringWithFormat:@"The version of header that is being created is unknown: %@", @(version)];
-            @throw [NSException exceptionWithName:NSInvalidArgumentException reason:reason userInfo:@{ @"requestedVersion": @(version) }];
+            @throw [NSException exceptionWithName:NSInvalidArgumentException reason:reason userInfo:@{ @"requestedVersion" : @(version) }];
+#else
+            return [[SDLV2ProtocolHeader alloc] initWithVersion:invalidSDLProtocolVersion];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProtocolHeader.m]: Would it be better to return `nil`? I'm not sure this is something we want to do in this PR. It's a change that would require some testing.

File: SmartDeviceLink/SDLProxy.m [hunk 8]
```diff
@@ -127,8 +124,8 @@ - (void)dispose {
 }
 
 - (void)dealloc {
-    [self destructObjects];
     [SDLDebugTool logInfo:@"SDLProxy Dealloc" withType:SDLDebugType_RPC toOutput:SDLDebugOutput_All toGroup:_debugConsoleGroupName];
+    [self destructObjects];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.m]: Why this change?

File: SmartDeviceLink/SDLProxy.m [hunk 9]
```diff
@@ -416,17 +413,19 @@ - (void)handleSystemRequest:(NSDictionary *)dict {
...
-    // Handle the various OnSystemRequest types
-    if (requestType == [SDLRequestType PROPRIETARY]) {
-        [self handleSystemRequestProprietary:systemRequest];
-    } else if (requestType == [SDLRequestType LOCK_SCREEN_ICON_URL]) {
-        [self handleSystemRequestLockScreenIconURL:systemRequest];
-    } else if (requestType == [SDLRequestType HTTP]) {
-        [self sdl_handleSystemRequestHTTP:systemRequest];
-    } else if (requestType == [SDLRequestType LAUNCH_APP]) {
-        [self sdl_handleSystemRequestLaunchApp:systemRequest];
+    if (systemRequest != nil) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.m]: This should never return nil. An assert would be better here. Have you seen this happen?

File: SmartDeviceLink/SDLProxy.m [hunk 10]
```diff
@@ -694,9 +693,11 @@ - (void)removeDelegate:(NSObject<SDLProxyListener> *)delegate {
 }
 
 - (void)invokeMethodOnDelegates:(SEL)aSelector withObject:(id)object {
+    __weak SDLProxy *weakSelf = self;
+    
     dispatch_async(dispatch_get_main_queue(), ^{
-        @autoreleasepool {
-            for (id<SDLProxyListener> listener in self.proxyListeners) {
+        if (!_alreadyDestructed){
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLProxy.m]: This should never be necessary, that whole process, including this variable, shouldn't exist. Have you seen errors based on this?

Why was `@autoreleasepool` removed? It should not have been.

File: SmartDeviceLink/SDLStreamDelegate.m [hunk 11]
```diff
@@ -89,4 +90,14 @@ - (void)stream:(NSStream *)stream handleEvent:(NSStreamEvent)eventCode {
     [SDLDebugTool logInfo:@"Stream Event End"];
 };
 
+- (void)dealloc{
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamDelegate.m]: Unnecessary, handled automatically by ARC.

File: SmartDeviceLink/SDLURLRequestTask.m [hunk 12]
```diff
@@ -16,7 +16,7 @@ @interface SDLURLRequestTask () <NSURLConnectionDelegate, NSURLConnectionDataDel
 
 @property (strong, nonatomic) NSURLConnection *connection;
 @property (strong, nonatomic, nullable) NSURLResponse *response;
-@property (copy, nonatomic) SDLURLConnectionRequestCompletionHandler completionHandler;
+@property (copy, nonatomic, nullable) SDLURLConnectionRequestCompletionHandler completionHandler;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLURLRequestTask.m]: Can't be nil, it's required in the only allowable `init`.

File: SmartDeviceLink/SDLURLRequestTask.m [hunk 13]
```diff
@@ -69,6 +69,7 @@ - (void)sdl_addData:(NSData *)data {
 #pragma mark - Cancel
 
 - (void)cancel {
+    self.completionHandler = nil;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLURLRequestTask.m]: Will cause the completion handler to not be called with the error.

**Comment by joeljfischer** [comment]
Closed for #597
